contracts/src/Functions.sol

```diff
@@ -65,7 +66,20 @@ library Functions {
             revert InvalidAmount();
         }
 
+        uint256 balanceBefore = IERC20(token).balanceOf(agent);
         IERC20(token).safeTransferFrom(sender, agent, amount);
+
+        uint256 balanceAfter = IERC20(token).balanceOf(agent);
+        if (balanceAfter < balanceBefore) {
+            revert InvalidAmount();
+        }
+
+        uint256 delta = balanceAfter - balanceBefore;
+        if (delta == 0 || delta > type(uint128).max) {
+            revert InvalidAmount();
+        }
```

Comment:
Is there a possibility of race conditions here? Ie the balance is altered between the before and after balance check, by some other transaction?

Reply:
Transactions are executed in order and single threaded so I do not think this can apply. There could be an issue of reentrancy within the token implementation or something like that that could do funky things but the before and after should pick that up.